packages/pg-v5/lib/psql.js

```diff
@@ -1,107 +1,247 @@
+  debug('Running sql file: %s', file.trim())
+
+  const childProcessOptions = {
+    encoding: 'utf8',
+    stdio: ['ignore', 'pipe', 'inherit']
+  }
+
+  const psqlArgs = ['-f', file, '--set', 'sslmode=require']
+
+  return {
+    dbEnv,
+    psqlArgs,
+    childProcessOptions,
+    pipeToStdout: true
+  }
+}
+
+function psqlInteractiveOptions (prompt, dbEnv) {
+  let psqlArgs = ['--set', `PROMPT1=${prompt}`, '--set', `PROMPT2=${prompt}`]
+  let psqlHistoryPath = process.env.HEROKU_PSQL_HISTORY
+  if (psqlHistoryPath) {
+    const fs = require('fs')
+    const path = require('path')
+    if (fs.existsSync(psqlHistoryPath) && fs.statSync(psqlHistoryPath).isDirectory()) {
+      let appLogFile = `${psqlHistoryPath}/${prompt.split(':')[0]}`
+      debug('Logging psql history to %s', appLogFile)
+      psqlArgs = psqlArgs.concat(['--set', `HISTFILE=${appLogFile}`])
+    } else if (fs.existsSync(path.dirname(psqlHistoryPath))) {
+      debug('Logging psql history to %s', psqlHistoryPath)
+      psqlArgs = psqlArgs.concat(['--set', `HISTFILE=${psqlHistoryPath}`])
     } else {
-      reject(err)
+      const cli = require('heroku-cli-util')
+      cli.warn(`HEROKU_PSQL_HISTORY is set but is not a valid path (${psqlHistoryPath})`)
     }
-  })
+  }
+  psqlArgs = psqlArgs.concat(['--set', 'sslmode=require'])
+
+  const childProcessOptions = {
+    stdio: 'inherit'
+  }
+
+  return {
+    dbEnv,
+    psqlArgs,
+    childProcessOptions
+  }
 }
 
-function execPsql (query, dbEnv) {
+function execPSQL ({ dbEnv, psqlArgs, childProcessOptions, pipeToStdout }) {
   const { spawn } = require('child_process')
-  return new Promise((resolve, reject) => {
-    let result = ''
-    debug('Running query: %s', query.trim())
-    let psql = spawn('psql', ['-c', query, '--set', 'sslmode=require'], { env: dbEnv, encoding: 'utf8', stdio: [ 'ignore', 'pipe', 'inherit' ] })
-    psql.stdout.on('data', function (data) {
-      result += data.toString()
-    })
-    psql.on('close', function (code) {
-      resolve(result)
-    })
-    handlePsqlError(reject, psql)
-  })
+
+  const options = {
+    env: dbEnv,
+    ...childProcessOptions
+  }
+
+  debug('opening psql process')
+  const psql = spawn('psql', psqlArgs, options)
+  psql.once('spawn', () => debug('psql process spawned'))
+
+  if (pipeToStdout) {
+    psql.stdout.pipe(process.stdout)
+  }
+
+  return psql
 }
 
-function execPsqlWithFile (file, dbEnv) {
-  const { spawn } = require('child_process')
-  return new Promise((resolve, reject) => {
-    let result = ''
-    debug('Running sql file: %s', file.trim())
-    let psql = spawn('psql', ['-f', file, '--set', 'sslmode=require'], { env: dbEnv, encoding: 'utf8', stdio: [ 'ignore', 'pipe', 'inherit' ] })
-    psql.stdout.on('data', function (data) {
-      result += data.toString()
-    })
-    psql.on('close', function (code) {
-      resolve(result)
+async function waitForPSQLExit (psql) {
+  try {
+    const exitCode = await once(psql, 'close')
```

Comment:
[events.once](https://nodejs.org/docs/latest-v14.x/api/events.html#events_events_once_emitter_name) will return a promise and resolve with the value of the event. If, while it is waiting for the event, an `error` event occurs, it will reject the promise. That's why you don't see the `.on('error')` event handler carried over in the new version.